contracts/vesting.token/src/vesting.token.cpp

```diff
@@ -0,0 +1,114 @@
+#include <vesting.token/vesting.token.hpp>
+
+namespace vestingtoken {
+
+    void vestingToken::updatedate(string sales_date_str, string launch_date_str) {
+        require_auth(get_self());
+        
+        vestingtoken::vesting_settings dates;
+        dates.sales_start_date = eosio::time_point_sec::from_iso_string(sales_date_str);
+        dates.launch_date = eosio::time_point_sec::from_iso_string(launch_date_str);
+
+        vestingtoken::settings launch_sales_dates_singleton(get_self(), get_self().value);
+        launch_sales_dates_singleton.set(dates, get_self());
+    }
+
+    void vestingToken::assigntokens(eosio::name holder, eosio::asset amount, int category_id) {
+        // Only the contract owner can call this function
+        require_auth(get_self());
+        
+        // Check if the provided category exists in the map
+        auto category_iter = vesting_categories.find(category_id);
+        eosio::check(category_iter != vesting_categories.end(), "Invalid vesting category");
+
+        // If the category exists, you can access it like this:
+        vesting_category category = category_iter->second;
+
+
+        // Check the symbol is correct and valid
+        auto sym = amount.symbol;
+        eosio::check(sym == system_resource_currency, "Symbol does not match system resource currency");
+        eosio::check(sym.precision() == system_resource_currency.precision(), "Symbol precision does not match");
+        eosio::check(amount.amount > 0, "Amount must be greater than 0");
+
+        // Create a new vesting schedule
+        vesting_allocations vesting_table(get_self(), holder.value);
+
+        // Check the number of rows for the user
+        uint32_t num_rows = 0;
+        for (auto itr = vesting_table.begin(); itr != vesting_table.end(); ++itr) {
+            ++num_rows;
+            if (num_rows > 20) {
+                eosio::check(false, "Cannot purchase tokens more than 20 times.");
```

Comment:
Add a comment to explain this briefly. And link to for more info
https://swcregistry.io/docs/SWC-128/

Just good to have documented as this is a pretty unintuitive code.